src/session_state.rs

```diff
@@ -11,25 +13,44 @@ use embedded_time::{
 /// The default duration to wait for a ping response from the broker.
 const PING_TIMEOUT: Seconds = Seconds(5);
 
-pub struct SessionState<Clock: embedded_time::Clock, const MSG_SIZE: usize, const MSG_COUNT: usize>
-{
+pub struct MessageRecord<const N: usize> {
+    pub msg: Vec<u8, N>,
+    transmitted: bool,
+    _id: u16,
+    qos: QoS,
+}
+
+pub struct SessionState<
+    TcpStack: TcpClientStack,
+    Clock: embedded_time::Clock,
+    const MSG_SIZE: usize,
+    const MSG_COUNT: usize,
+> {
     keep_alive_interval: Option<Milliseconds<u32>>,
     ping_timeout: Option<Instant<Clock>>,
     next_ping: Option<Instant<Clock>>,
     pub broker: SocketAddr,
     pub maximum_packet_size: Option<u32>,
     pub client_id: String<64>,
     pub pending_subscriptions: Vec<u16, 32>,
-    pub pending_publish: LinearMap<u16, Vec<u8, MSG_SIZE>, MSG_COUNT>,
-    pub pending_publish_ordering: Vec<u16, MSG_COUNT>,
+    pending_publish: LinearMap<u16, MessageRecord<MSG_SIZE>, MSG_COUNT>,
+    pending_publish_ordering: Vec<u16, MSG_COUNT>,
```

Comment:
I was thinking about using a `heapless::Pool`, but was a bit concerned about the portability caveat: https://docs.rs/heapless/0.7.16/heapless/pool/index.html#portability

Similarly, the sorted_linked_list is a _sorted_ list that reorders items as they are added.

What would be ideal is just an unsorted linked list with static storage backing it. I'm somewhat unsure why `heapless` doesn't have this to begin with.

Reply:
`heapless::Deque` seems to _almost_ do what we want, but it doesn't allow indexing into the Deque. I'll make a PR to add this support, since it seems like a nice change to make the Deque into a ring-buffer.

Reply:
You are also sorting `pending_publish_ordering`. Why would it be worse then sorted_linked_list where sorting is just as cheap?

Reply:
We are not sorting that list. It's explicitly a FIFO. MQTTv5's spec requires that the republication order match the original publication order.

Reply:
Also, a Deque doesn't work because we don't know the order that we will receive the PUBACK messages, so we may need to rearrange the ordering.

The spec doesn't appear to place any restriction on PUBACK ordering from the SERVER, only from the client.

Reply:
I didn't find where you are enforcing that FIFO behavior. You are removing from where you find it, right?

Then `spsc::Queue()`!

Reply:
`Vec::remove` maintains the order of the vector by shifting all items to the left.

Spsc also cannot be used because the queue order may need to be reordered as we receive PUBACKs

Reply:
You are maintaining but not enforcing it (re incoming packets). I.e. you don't always remove from the head.

Reply:
Correct - the spec doesn't state that order has to be maintained w.r.t incoming packets from the server, but it _does_ maintain that we have to republish in the original order of publication.

Reply:
Then `sorted_linked_list` where all items compare `Equal` would probably work just fine.

Reply:
I.e. just a linked list.

Reply:
sorted_linked_list doesn't appear to expose an `iter_mut()` method, so we can't iterate across all of the elements.

I think the best approach is to use the heapless::pool. I think it has an atomic-polyfill for e.g. thumbv6 targets, and we can cross that bridge if we come to it later. All of our intended targets support atomics.

Reply:
Hmm... even `heapless::Pool` requires that some memory be borrowed to the `SessionState` (e.g. 'static), which would require us to:
1. Template the `Minimq` type with the lifetime, pulling it all the way through all types
2. User has to pass in a [u8] buffer, which may or may not be large enough for even a single packet

This doesn't seem super appealing to me either, although it could actually relax memory constraints for users that do not want any QoS over AtMostOnce, since they could specify a buffer size of zero.

Reply:
I think the best approach here may be to leave it alone for now and open an issue to address refactoring in the future. It looks like none of the typical solutions can offer us a desirable data structure right now.

Reply:
Agreed!

Reply:
Spawned https://github.com/quartiq/minimq/issues/92